File: packages/react-native-nitro-shaders/ios/HybridNitroShaders.swift
//
//  HybridNitroShaders.swift
//  Pods
//
//  Created by Giulio Amato on 03/07/2026.
//
//  iOS runtime of the material orb. Mirrors the Android runtime structure:
//  the orb fragment program is assembled per material from bundled MSL sources
//  (orb-core.msl + material-<name>.msl + orb-main.msl) and compiled at runtime,
//  one cached pipeline per material. Environments are equirectangular PNGs in
//  the NitroShaders resource bundle (same assets as Android).
//
//  First porting pass: the silhouette is cut in-shader; the contact shadow of
//  the Android Skin is NOT rendered yet (declared debt).
//

import Foundation
import Metal
import MetalKit
import UIKit

// MTKViewDelegate requires NSObjectProtocol while the Nitro base class is not an
// NSObject — a small proxy forwards the delegate callbacks to the hybrid object.
private final class OrbViewDelegate: NSObject, MTKViewDelegate {
  weak var owner: HybridNitroShaders?

  func mtkView(_ view: MTKView, drawableSizeWillChange size: CGSize) {
    owner?.drawableSizeWillChange()
  }

  func draw(in view: MTKView) {
    owner?.render(in: view)
  }
}

class HybridNitroShaders : HybridNitroShadersSpec {
  // Legacy "solid" pipeline uniforms.
  private struct SolidUniforms {
    var color: SIMD4<Float>
    var time: Float
  }

  // MUST mirror OrbUniforms in ios/Shaders/orb-core.msl (same field order —
  // float4 first, then float2 pair, then scalar floats).
  private struct OrbUniforms {
    var lightDir: SIMD4<Float>
    var resolution: SIMD2<Float>
    var envSize: SIMD2<Float>
    var time: Float
    var speed: Float
    var morph: Float
    var orbit: Float
    var pattern: Float
    var patternScale: Float
    var patternDistortion: Float
    var tint: Float
    var opacity: Float
    var envRot: Float
    var hdrBoost: Float
    var density: Float
    var smoothness: Float
    var silDensity: Float
    var smoothSil: Float
    var motionSeed: Float
  }

  private let metalView: MTKView
  private let viewDelegate = OrbViewDelegate()
  private let device: MTLDevice?
  private let commandQueue: MTLCommandQueue?
  private var solidLibrary: MTLLibrary?
  private var solidPipeline: MTLRenderPipelineState?
  // One assembled+compiled pipeline per orb material, built lazily.
  private var orbPipelines: [String: MTLRenderPipelineState] = [:]
  // Environment textures by asset name ("studio", "lab-3", ...), lazy.
  private var envTextures: [String: MTLTexture] = [:]
  private var startTime = CACurrentMediaTime()

  var view: UIView {
    metalView
  }

  // ---- Surface selection + lifecycle ----

  var shader: String = "solid" {
    didSet { drawIfNeeded() }
  }

  var color: String = "#000000" {
    didSet { drawIfNeeded() }
  }

  var colors: [String] = [] {
    didSet { drawIfNeeded() }
  }

  var animated: Bool = false {
    didSet { applyAnimationState() }
  }

  var paused: Bool = false {
    didSet { applyAnimationState() }
  }

  var debugTime: Double = -1 {
    didSet {
      applyAnimationState()
      drawIfNeeded()
    }
  }

  // ---- Material orb: semantic parameter set ----

  var material: String = "metal" {
    didSet { drawIfNeeded() }
  }

  var speed: Double = 1 {
    didSet { drawIfNeeded() }
  }

  var morph: Double = 0.5 {
    didSet { drawIfNeeded() }
  }

  var orbit: Double = 1 {
    didSet { drawIfNeeded() }
  }

  var pattern: String = "auto" {
    didSet { drawIfNeeded() }
  }

  var patternScale: Double = 1 {
    didSet { drawIfNeeded() }
  }

  var patternDistortion: Double = 0.5 {
    didSet { drawIfNeeded() }
  }

  var tint: Double = 0.5 {
    didSet { drawIfNeeded() }
  }

  var opacity: Double = 1 {
    didSet { drawIfNeeded() }
  }

  var lightAzimuth: Double = 0 {
    didSet { drawIfNeeded() }
  }

  var lightElevation: Double = Double.pi / 2 {
    didSet { drawIfNeeded() }
  }

  var environment: Double = -1 {
    didSet { drawIfNeeded() }
  }

  var envRotation: Double = 0 {
    didSet { drawIfNeeded() }
  }

  var hdr: Bool = false {
    didSet { drawIfNeeded() }
  }

  var density: Double = 1 {
    didSet { drawIfNeeded() }
  }

  var smoothness: Double = 0 {
    didSet { drawIfNeeded() }
  }

  // ---- Motion uniform contract ----

  var motionType: Double = 0 {
    didSet { drawIfNeeded() }
  }

  var motionSpeed: Double = 0 {
    didSet { drawIfNeeded() }
  }

  var motionAmp: Double = 0 {
    didSet { drawIfNeeded() }
  }

  var motionWarp: Double = 0 {
    didSet { drawIfNeeded() }
  }

  var motionDetail: Double = 0 {
    didSet { drawIfNeeded() }
  }

  var motionSeed: Double = 0 {
    didSet { drawIfNeeded() }
  }

  var motionPeriod: Double = 0 {
    didSet { drawIfNeeded() }
  }

  override init() {
    let metalDevice = MTLCreateSystemDefaultDevice()
    device = metalDevice
    commandQueue = metalDevice?.makeCommandQueue()
    metalView = MTKView(frame: .zero, device: metalDevice)
    super.init()

    // Transparent surface: the orb floats over whatever is behind the view.
    metalView.clearColor = MTLClearColor(red: 0, green: 0, blue: 0, alpha: 0)
    metalView.isOpaque = false
    metalView.backgroundColor = .clear
    metalView.framebufferOnly = true
    metalView.isPaused = true
    metalView.enableSetNeedsDisplay = true
    viewDelegate.owner = self
    metalView.delegate = viewDelegate
    solidLibrary = try? metalDevice?.makeLibrary(source: Self.solidShaderSource, options: nil)
  }

  fileprivate func drawableSizeWillChange() {
    drawIfNeeded()
  }

  fileprivate func render(in view: MTKView) {
    guard
      let commandQueue,
      let drawable = view.currentDrawable,
      let descriptor = view.currentRenderPassDescriptor
    else {
      return
    }

    let time = Float(debugTime >= 0 ? debugTime : CACurrentMediaTime() - startTime)

    let commandBuffer = commandQueue.makeCommandBuffer()
    let encoder = commandBuffer?.makeRenderCommandEncoder(descriptor: descriptor)

    if shader == "materialOrb" {
      drawOrb(encoder: encoder, time: time)
    } else {
      drawSolid(encoder: encoder, time: time)
    }

    encoder?.endEncoding()
    commandBuffer?.present(drawable)
    commandBuffer?.commit()
  }

  // ---- Material orb drawing ----

  private static let orbMaterials: Set<String> = ["metal", "water", "iridescent", "glass"]

  private var resolvedMaterial: String {
    Self.orbMaterials.contains(material) ? material : "metal"
  }

  // Motion params fall back to the semantic params when unset (0) — mirrors
  // ShaderSurfaceView.resolvedMotion* on Android.
  private var resolvedSpeed: Float { motionSpeed > 0 ? Float(motionSpeed) : Float(speed) }
  private var resolvedMorph: Float { motionAmp > 0 ? Float(motionAmp) : Float(morph) }
  private var resolvedWarp: Float { motionWarp > 0 ? Float(motionWarp) : Float(patternDistortion) }
  private var resolvedScale: Float { motionDetail > 0 ? Float(motionDetail) : Float(patternScale) }

  private func defaultPatternIndex(for material: String) -> Float {
    switch material {
    case "water": return 2   // ripples
    case "glass": return 1   // bands
    default: return 0        // folds (metal/iridescent)
    }
  }

  private func patternIndex(for material: String) -> Float {
    switch pattern {
    case "folds": return 0
    case "bands": return 1
    case "ripples": return 2
    default: return defaultPatternIndex(for: material)
    }
  }

  // Silhouette family + density + default env, mirroring ShaderMaterial.kt.
  private func isSmoothSilhouette(_ material: String) -> Bool {
    material == "water" || material == "glass"
  }

  private func silhouetteDensity(_ material: String) -> Float {
    switch material {
    case "metal": return 0.92
    case "water", "glass": return 0.45
    default: return 0.62
    }
  }

  private func defaultEnvName(_ material: String) -> String {
    switch material {
    case "water": return "water-env"
    case "glass": return "glass-env"
    default: return "studio"
    }
  }

  private func drawOrb(encoder: MTLRenderCommandEncoder?, time: Float) {
    let material = resolvedMaterial
    guard
      let pipeline = orbPipeline(for: material),
      let envTexture = envTexture(for: material)
    else {
      return
    }

    let size = metalView.drawableSize
    let az = Float(lightAzimuth)
    let el = Float(lightElevation)
    var uniforms = OrbUniforms(
      lightDir: SIMD4<Float>(sin(az) * cos(el), sin(el), cos(az) * cos(el), 0),
      resolution: SIMD2<Float>(Float(size.width), Float(size.height)),
      envSize: SIMD2<Float>(Float(envTexture.width), Float(envTexture.height)),
      time: time,
      speed: resolvedSpeed,
      morph: resolvedMorph,
      orbit: Float(orbit),
      pattern: patternIndex(for: material),
      patternScale: resolvedScale,
      patternDistortion: resolvedWarp,
      tint: Float(tint),
      opacity: Float(opacity),
      envRot: Float(envRotation),
      hdrBoost: hdr ? 1 : 0,
      density: Float(density),
      smoothness: Float(smoothness),
      silDensity: silhouetteDensity(material),
      smoothSil: isSmoothSilhouette(material) ? 1 : 0,
      motionSeed: Float(motionSeed)
    )

    encoder?.setRenderPipelineState(pipeline)
    encoder?.setFragmentBytes(&uniforms, length: MemoryLayout<OrbUniforms>.stride, index: 0)
    encoder?.setFragmentTexture(envTexture, index: 0)
    encoder?.drawPrimitives(type: .triangle, vertexStart: 0, vertexCount: 3)
  }

  private func drawSolid(encoder: MTLRenderCommandEncoder?, time: Float) {
    guard let pipeline = solidPipelineState() else {
      return
    }
    var uniforms = SolidUniforms(color: Self.parseColor(color), time: time)
    encoder?.setRenderPipelineState(pipeline)
    encoder?.setFragmentBytes(&uniforms, length: MemoryLayout<SolidUniforms>.stride, index: 0)
    encoder?.drawPrimitives(type: .triangle, vertexStart: 0, vertexCount: 3)
  }

  // ---- Pipelines + resources ----

  private static let resourceBundle: Bundle? = {
    let frameworkBundle = Bundle(for: HybridNitroShaders.self)
    for candidate in [frameworkBundle, Bundle.main] {
      if let url = candidate.url(forResource: "NitroShaders", withExtension: "bundle"),
         let bundle = Bundle(url: url) {
        return bundle
      }
    }
    return nil
  }()

  private func loadShaderSource(_ name: String) -> String? {
    guard let url = Self.resourceBundle?.url(forResource: name, withExtension: "msl") else {
      return nil
    }
    return try? String(contentsOf: url, encoding: .utf8)
  }

  private func orbPipeline(for material: String) -> MTLRenderPipelineState? {
    if let cached = orbPipelines[material] {
      return cached
    }
    guard
      let device,
      let core = loadShaderSource("orb-core"),
      let mat = loadShaderSource("material-\(material)"),
      let main = loadShaderSource("orb-main")
    else {
      return nil
    }
    do {
      let source = core + "\n" + mat + "\n" + main
      let library = try device.makeLibrary(source: source, options: nil)
      let descriptor = MTLRenderPipelineDescriptor()
      descriptor.vertexFunction = library.makeFunction(name: "orbVertex")
      descriptor.fragmentFunction = library.makeFunction(name: "orbFragment")
      let attachment = descriptor.colorAttachments[0]
      attachment?.pixelFormat = metalView.colorPixelFormat
      // The fragment returns premultiplied alpha.
      attachment?.isBlendingEnabled = true
      attachment?.sourceRGBBlendFactor = .one
      attachment?.destinationRGBBlendFactor = .oneMinusSourceAlpha
      attachment?.sourceAlphaBlendFactor = .one
      attachment?.destinationAlphaBlendFactor = .oneMinusSourceAlpha
      let state = try device.makeRenderPipelineState(descriptor: descriptor)
      orbPipelines[material] = state
      return state
    } catch {
      return nil
    }
  }

  private func envTexture(for material: String) -> MTLTexture? {
    let labIndex = Int(environment)
    let name = labIndex >= 0 ? "lab-\(labIndex)" : defaultEnvName(material)
    if let cached = envTextures[name] {
      return cached
    }
    guard
      let device,
      let url = Self.resourceBundle?.url(forResource: name, withExtension: "png")
    else {
      return nil
    }
    let loader = MTKTextureLoader(device: device)
    let options: [MTKTextureLoader.Option: Any] = [
      .SRGB: false,
      .textureUsage: MTLTextureUsage.shaderRead.rawValue,
      .textureStorageMode: MTLStorageMode.private.rawValue,
    ]
    guard let texture = try? loader.newTexture(URL: url, options: options) else {
      return nil
    }
    envTextures[name] = texture
    return texture
  }

  private func solidPipelineState() -> MTLRenderPipelineState? {
    if let cached = solidPipeline {
      return cached
    }
    guard let device, let library = solidLibrary else {
      return nil
    }
    do {
      let descriptor = MTLRenderPipelineDescriptor()
      descriptor.vertexFunction = library.makeFunction(name: "vertexMain")
      descriptor.fragmentFunction = library.makeFunction(name: "fragmentMain")
      descriptor.colorAttachments[0].pixelFormat = metalView.colorPixelFormat
      let state = try device.makeRenderPipelineState(descriptor: descriptor)
      solidPipeline = state
      return state
    } catch {
      return nil
    }
  }

  // ---- Animation / invalidation ----

  private var shouldAnimate: Bool {
    animated && !paused && debugTime < 0
  }

  private func applyAnimationState() {
    metalView.enableSetNeedsDisplay = !shouldAnimate
    metalView.isPaused = !shouldAnimate
    drawIfNeeded()
  }

  private func drawIfNeeded() {
    if !shouldAnimate {
      metalView.setNeedsDisplay()
    }
  }

  // ---- Helpers ----

  private static func parseColor(_ value: String) -> SIMD4<Float> {
    let hex = value.trimmingCharacters(in: CharacterSet(charactersIn: "#"))
    guard hex.count == 6 || hex.count == 8, let raw = UInt32(hex, radix: 16) else {
      return SIMD4<Float>(0, 0, 0, 1)
    }

    if hex.count == 8 {
      return SIMD4<Float>(
        Float((raw >> 24) & 0xff) / 255,
        Float((raw >> 16) & 0xff) / 255,
        Float((raw >> 8) & 0xff) / 255,
        Float(raw & 0xff) / 255
      )
    }

    return SIMD4<Float>(
      Float((raw >> 16) & 0xff) / 255,
      Float((raw >> 8) & 0xff) / 255,
      Float(raw & 0xff) / 255,
      1
    )
  }

  private static let solidShaderSource = """
  #include <metal_stdlib>
  using namespace metal;

  struct SolidVertexOut {
    float4 position [[position]];
  };

  struct SolidUniforms {
    float4 color;
    float time;
  };

  vertex SolidVertexOut vertexMain(uint vertexID [[vertex_id]]) {
    float2 positions[3] = {
      float2(-1.0, -1.0),
      float2(3.0, -1.0),
      float2(-1.0, 3.0)
    };

    SolidVertexOut out;
    out.position = float4(positions[vertexID], 0.0, 1.0);
    return out;
  }

  fragment float4 fragmentMain(SolidVertexOut in [[stage_in]],
                               constant SolidUniforms& uniforms [[buffer(0)]]) {
    return uniforms.color;
  }

  """
}

File: packages/react-native-nitro-shaders/android/src/main/java/com/nitroshaders/HybridNitroShaders.kt
package com.margelo.nitro.nitroshaders

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.view.Choreographer
import android.view.View
import androidx.annotation.Keep
import com.facebook.proguard.annotations.DoNotStrip
import com.facebook.react.uimanager.ThemedReactContext

@Keep
@DoNotStrip
class HybridNitroShaders(val context: ThemedReactContext): HybridNitroShadersSpec() {
    override val view: View = ShaderSurfaceView(context)

    private val shaderSurfaceView: ShaderSurfaceView
        get() = view as ShaderSurfaceView

    // ---- Surface selection + lifecycle ----

    private var _shader = "solid"
    override var shader: String
        get() = _shader
        set(value) {
            _shader = value
            shaderSurfaceView.shader = value
        }

    private var _color = "#000000"
    override var color: String
        get() = _color
        set(value) {
            _color = value
            shaderSurfaceView.color = value
        }

    private var _colors: Array<String> = arrayOf()
    override var colors: Array<String>
        get() = _colors
        set(value) {
            _colors = value
            shaderSurfaceView.colors = value
        }

    private var _animated = false
    override var animated: Boolean
        get() = _animated
        set(value) {
            _animated = value
            shaderSurfaceView.animated = value
        }

    private var _paused = false
    override var paused: Boolean
        get() = _paused
        set(value) {
            _paused = value
            shaderSurfaceView.paused = value
        }

    private var _debugTime = -1.0
    override var debugTime: Double
        get() = _debugTime
        set(value) {
            _debugTime = value
            shaderSurfaceView.debugTime = value
        }

    // ---- Material orb: semantic parameter set ----

    private var _material = "metal"
    override var material: String
        get() = _material
        set(value) {
            _material = value
            shaderSurfaceView.material = value
        }

    private var _speed = 1.0
    override var speed: Double
        get() = _speed
        set(value) {
            _speed = value
            shaderSurfaceView.speed = value
        }

    private var _morph = 0.5
    override var morph: Double
        get() = _morph
        set(value) {
            _morph = value
            shaderSurfaceView.morph = value
        }

    private var _orbit = 1.0
    override var orbit: Double
        get() = _orbit
        set(value) {
            _orbit = value
            shaderSurfaceView.orbit = value
        }

    private var _pattern = "auto"
    override var pattern: String
        get() = _pattern
        set(value) {
            _pattern = value
            shaderSurfaceView.pattern = value
        }

    private var _patternScale = 1.0
    override var patternScale: Double
        get() = _patternScale
        set(value) {
            _patternScale = value
            shaderSurfaceView.patternScale = value
        }

    private var _patternDistortion = 0.5
    override var patternDistortion: Double
        get() = _patternDistortion
        set(value) {
            _patternDistortion = value
            shaderSurfaceView.patternDistortion = value
        }

    private var _tint = 0.5
    override var tint: Double
        get() = _tint
        set(value) {
            _tint = value
            shaderSurfaceView.tint = value
        }

    private var _opacity = 1.0
    override var opacity: Double
        get() = _opacity
        set(value) {
            _opacity = value
            shaderSurfaceView.opacity = value
        }

    private var _lightAzimuth = 0.0
    override var lightAzimuth: Double
        get() = _lightAzimuth
        set(value) {
            _lightAzimuth = value
            shaderSurfaceView.lightAzimuth = value
        }

    private var _lightElevation = Math.PI / 2
    override var lightElevation: Double
        get() = _lightElevation
        set(value) {
            _lightElevation = value
            shaderSurfaceView.lightElevation = value
        }

    private var _environment = -1.0
    override var environment: Double
        get() = _environment
        set(value) {
            _environment = value
            shaderSurfaceView.environment = value
        }

    private var _envRotation = 0.0
    override var envRotation: Double
        get() = _envRotation
        set(value) {
            _envRotation = value
            shaderSurfaceView.envRotation = value
        }

    private var _hdr = false
    override var hdr: Boolean
        get() = _hdr
        set(value) {
            _hdr = value
            shaderSurfaceView.hdr = value
        }

    private var _density = 1.0
    override var density: Double
        get() = _density
        set(value) {
            _density = value
            shaderSurfaceView.density = value
        }

    private var _smoothness = 0.0
    override var smoothness: Double
        get() = _smoothness
        set(value) {
            _smoothness = value
            shaderSurfaceView.smoothness = value
        }

    // ---- Motion uniform contract ----

    private var _motionType = 0.0
    override var motionType: Double
        get() = _motionType
        set(value) {
            _motionType = value
            shaderSurfaceView.motionType = value
        }

    private var _motionSpeed = 0.0
    override var motionSpeed: Double
        get() = _motionSpeed
        set(value) {
            _motionSpeed = value
            shaderSurfaceView.motionSpeed = value
        }

    private var _motionAmp = 0.0
    override var motionAmp: Double
        get() = _motionAmp
        set(value) {
            _motionAmp = value
            shaderSurfaceView.motionAmp = value
        }

    private var _motionWarp = 0.0
    override var motionWarp: Double
        get() = _motionWarp
        set(value) {
            _motionWarp = value
            shaderSurfaceView.motionWarp = value
        }

    private var _motionDetail = 0.0
    override var motionDetail: Double
        get() = _motionDetail
        set(value) {
            _motionDetail = value
            shaderSurfaceView.motionDetail = value
        }

    private var _motionSeed = 0.0
    override var motionSeed: Double
        get() = _motionSeed
        set(value) {
            _motionSeed = value
            shaderSurfaceView.motionSeed = value
        }

    private var _motionPeriod = 0.0
    override var motionPeriod: Double
        get() = _motionPeriod
        set(value) {
            _motionPeriod = value
            shaderSurfaceView.motionPeriod = value
        }
}

internal class ShaderSurfaceView(context: Context): View(context), Choreographer.FrameCallback {
    // ---- Surface ----
    var shader: String = "solid"
        set(value) {
            field = value
            invalidate()
        }

    var color: String = "#000000"
        set(value) {
            field = value
            parsedColor = parseColor(value)
            invalidate()
        }

    var colors: Array<String> = arrayOf()
        set(value) {
            field = value
            parsedColors = parseColors(value)
            invalidate()
        }

    var animated: Boolean = false
        set(value) {
            field = value
            updateFrameCallback()
        }

    var paused: Boolean = false
        set(value) {
            field = value
            updateFrameCallback()
        }

    var debugTime: Double = -1.0
        set(value) {
            field = value
            updateFrameCallback()
            invalidate()
        }

    // ---- Material orb semantic params ----
    var material: String = "metal"
        set(value) {
            field = value
            invalidate()
        }

    var speed: Double = 1.0
        set(value) {
            field = value
            invalidate()
        }

    var morph: Double = 0.5
        set(value) {
            field = value
            invalidate()
        }

    var orbit: Double = 1.0
        set(value) {
            field = value
            invalidate()
        }

    var pattern: String = "auto"
        set(value) {
            field = value
            invalidate()
        }

    var patternScale: Double = 1.0
        set(value) {
            field = value
            invalidate()
        }

    var patternDistortion: Double = 0.5
        set(value) {
            field = value
            invalidate()
        }

    var tint: Double = 0.5
        set(value) {
            field = value
            invalidate()
        }

    var opacity: Double = 1.0
        set(value) {
            field = value
            invalidate()
        }

    var lightAzimuth: Double = 0.0
        set(value) {
            field = value
            invalidate()
        }

    var lightElevation: Double = Math.PI / 2
        set(value) {
            field = value
            invalidate()
        }

    var environment: Double = -1.0
        set(value) {
            field = value
            invalidate()
        }

    var envRotation: Double = 0.0
        set(value) {
            field = value
            invalidate()
        }

    var hdr: Boolean = false
        set(value) {
            field = value
            invalidate()
        }

    var density: Double = 1.0
        set(value) {
            field = value
            invalidate()
        }

    var smoothness: Double = 0.0
        set(value) {
            field = value
            invalidate()
        }

    // ---- Motion ----
    var motionType: Double = 0.0
        set(value) {
            field = value
            invalidate()
        }

    var motionSpeed: Double = 0.0
        set(value) {
            field = value
            invalidate()
        }

    var motionAmp: Double = 0.0
        set(value) {
            field = value
            invalidate()
        }

    var motionWarp: Double = 0.0
        set(value) {
            field = value
            invalidate()
        }

    var motionDetail: Double = 0.0
        set(value) {
            field = value
            invalidate()
        }

    var motionSeed: Double = 0.0
        set(value) {
            field = value
            invalidate()
        }

    var motionPeriod: Double = 0.0
        set(value) {
            field = value
            invalidate()
        }

    internal val appContext = context.applicationContext
    internal var parsedColor = Color.BLACK
        private set
    internal var parsedColors: FloatArray = FloatArray(0)
        private set
    private var frameCallbackPosted = false
    private val startTimeNanos = System.nanoTime()

    // Registry of shader materials (Strategy pattern). onDraw looks up the active
    // material by [shader] name and delegates rendering to it; unknown names fall
    // back to the default "solid" material.
    private val solidMaterial: ShaderMaterial = SolidMaterial()
    private val materials: Map<String, ShaderMaterial> = listOf(
        MaterialOrbMaterial(),
        solidMaterial
    ).associateBy { it.name }

    override fun onAttachedToWindow() {
        super.onAttachedToWindow()
        updateFrameCallback()
    }

    override fun onDetachedFromWindow() {
        stopFrameCallback()
        super.onDetachedFromWindow()
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        val w = width.toFloat()
        val h = height.toFloat()

        val activeMaterial = materials[shader] ?: solidMaterial
        if (activeMaterial.isAvailable()) {
            activeMaterial.draw(canvas, this, w, h)
        } else {
            activeMaterial.drawFallback(canvas, this, w, h)
        }
    }

    // Motion params fall back to the semantic params when unset (0) — the motion
    // system can override any of them at runtime without touching the others.
    internal fun resolvedMotionSpeed(): Float {
        return if (motionSpeed > 0.0) motionSpeed.toFloat() else speed.toFloat()
    }

    internal fun resolvedMotionAmp(): Float {
        return if (motionAmp > 0.0) motionAmp.toFloat() else morph.toFloat()
    }

    internal fun resolvedMotionWarp(): Float {
        return if (motionWarp > 0.0) motionWarp.toFloat() else patternDistortion.toFloat()
    }

    internal fun resolvedMotionDetail(): Float {
        return if (motionDetail > 0.0) motionDetail.toFloat() else patternScale.toFloat()
    }

    override fun doFrame(frameTimeNanos: Long) {
        frameCallbackPosted = false
        invalidate()
        updateFrameCallback()
    }

    private fun updateFrameCallback() {
        if (animated && !paused && debugTime < 0 && isAttachedToWindow) {
            if (!frameCallbackPosted) {
                frameCallbackPosted = true
                Choreographer.getInstance().postFrameCallback(this)
            }
        } else {
            stopFrameCallback()
        }
    }

    private fun stopFrameCallback() {
        if (frameCallbackPosted) {
            Choreographer.getInstance().removeFrameCallback(this)
            frameCallbackPosted = false
        }
    }

    internal fun currentTimeSeconds(): Float {
        if (debugTime >= 0) {
            return debugTime.toFloat()
        }

        return ((System.nanoTime() - startTimeNanos).toDouble() / 1_000_000_000.0).toFloat()
    }

    internal fun parseColor(value: String): Int {
        return try {
            Color.parseColor(value)
        } catch (_: IllegalArgumentException) {
            Color.BLACK
        }
    }

    // Parse up to 6 hex colors into a flat [r,g,b, ...] float array (0..1). Extras ignored.
    private fun parseColors(values: Array<String>): FloatArray {
        val count = values.size.coerceAtMost(6)
        val out = FloatArray(count * 3)
        for (i in 0 until count) {
            val c = parseColor(values[i])
            out[i * 3] = Color.red(c) / 255f
            out[i * 3 + 1] = Color.green(c) / 255f
            out[i * 3 + 2] = Color.blue(c) / 255f
        }
        return out
    }

    internal fun red(): Float = Color.red(parsedColor) / 255f
    internal fun green(): Float = Color.green(parsedColor) / 255f
    internal fun blue(): Float = Color.blue(parsedColor) / 255f
    internal fun alpha(): Float = Color.alpha(parsedColor) / 255f

}
